packages/react-provider/src/index.tsx

```diff
@@ -33,155 +33,159 @@ export type {
+        });
+      }
+    }, [_transport, clientKey, wsOptions, userSignature]);
 
-    actions.init({
+    const [state, dispatch] = useReducer(reducer, {
       apiUrl,
+      brand,
       brandId,
       clientKey,
       transport,
       userId,
       userSignature,
+      middleware,
     });
-  }, [apiUrl, clientKey, transport, userId, userSignature, brandId]);
-
-  useEffect(() => {
-    if (brand) {
-      // if we pass in brand, don't fetch it
-      return;
-    }
-
-    actions.getBrand(brandId);
-  }, [brand, brandId]);
-
-  useEffect(() => {
-    if (!state.brand || !clientKey || !userId) {
-      return;
-    }
-
-    localStorage.setItem(
-      `${clientKey}/${userId}/provider`,
-      JSON.stringify({
-        brand: state.brand,
-      })
-    );
-  }, [state.brand, clientKey, userId]);
 
-  useEffect(() => {
-    if (!clientKey || !userId) {
-      return;
-    }
+    const actions = useCourierActions(state, dispatch);
 
-    const localStorageState = localStorage.getItem(
-      `${clientKey}/${userId}/provider`
-    );
+    useEffect(() => {
+      if (_transport) {
+        // this means the transport was passed in and we shouldn't subscribe
+        return;
+      }
 
-    if (localStorageState) {
-      try {
-        const { brand } = JSON.parse(localStorageState);
-        actions.setBrand(brand);
-      } catch (ex) {
-        console.log("error", ex);
+      if (!transport || !userId) {
+        return;
       }
-    }
-  }, [clientKey, userId]);
-
-  return (
-    <CourierContext.Provider
-      value={{
-        ...(state as any),
-        ...actions,
-        dispatch,
-      }}
-    >
-      {children}
-    </CourierContext.Provider>
-  );
-};
+
+      const courierTransport = transport as CourierTransport;
+      courierTransport.subscribe(userId);
+
+      if (onMessage) {
+        courierTransport.intercept(onMessage);
+      }
+
+      courierTransport.listen({
+        id: "deliver-tracking",
+        listener: (courierEvent) => {
+          const courierData = courierEvent?.data?.data;
+          if (!courierData?.trackingIds?.deliverTrackingId) {
+            return;
+          }
+
+          actions.createTrackEvent(courierData?.trackingIds?.deliverTrackingId);
+        },
+      });
+
+      return () => {
+        courierTransport.unsubscribe(userId);
+      };
+    }, [transport, userId]);
+
+    useEffect(() => {
+      if (!_transport && (!clientKey || !userId)) {
+        return;
+      }
+
+      actions.init({
+        apiUrl,
+        brandId,
+        clientKey,
+        transport,
+        userId,
+        userSignature,
+      });
+    }, [apiUrl, clientKey, transport, userId, userSignature, brandId]);
+
+    useEffect(() => {
+      if (brand) {
+        // if we pass in brand, don't fetch it
+        return;
+      }
+
+      if (!clientKey || !userId) {
+        return;
+      }
+
+      actions.getBrand(brandId);
+    }, [actions, brand, brandId, clientKey, userId]);
```

Comment:
this useEffect is all that changed, i added conditional on line 141 and updated the deps in 146